SCUMM HE: Use Miles AdLib driver

nukeykt opened the pull request:
This patch replaces generic AdLib driver with Miles AdLib driver for Humongous games. This resolves 7880 and 6788 tickets.

wjp commented:
Thank you for the contribution!

Could you give a more in-depth explanation? Both the pull request and the commit message are lacking any information. (Also, there seem to be a few unrelated changes in there?)

nukeykt commented:
Ok.
I've added special AdLib player. This player extracts FM timbres from DRV resource file and opens Miles driver.
I've also made some changes to music APIs. Humongous games pass two parameters when starting song: sound id and track id.
Also i've added master volume control and fixed pitchbend bug in Miles driver.
The final change is Putt Putt Goes to the Moon Demo tweak. PuttMoon unlike other games has GM music, so it uses different timbre file, but its demo version reuses Putt Putt Parade's music. To resolve this I've just changed demo's game id to generic HEGAME.

sev- commented on audio/miles_adlib.cpp:
```diff
@@ -723,7 +726,7 @@ void MidiDriver_Miles_AdLib::updatePhysicalFmVoice(byte virtualFmVoice, bool key
 	if (registerUpdateFlags & kMilesAdLibUpdateFlags_Reg_40) {
 		// Calculate new volume
-		byte midiVolume = _midiChannels[midiChannel].currentVolume;
+		byte midiVolume = (_midiChannels[midiChannel].currentVolume * _masterVolume) >> 8;
```
Why are you changing here the behaviour of the driver? The rest of the driver users are not aware of this change, and still pass volume as a byte, and will effectively mute sound now.

nukeykt replied:
AFAIK all other games handle volume control outside of this module. We defaulting this value to 255, so other games should work fine.

sev- commented on engines/scumm/players/player_he.cpp:
```diff
@@ -0,0 +1,194 @@
+	if (!file.open(drvName))
+		error("Player_HE::loadAdLibBank(): could not open %s", drvName.c_str());
+
+	uint32 size = uint32(file.size());
```
Is it a type casting? Please use canonical way: (uint32)file.size();

sev- commented on engines/scumm/players/player_he.cpp:
```diff
@@ -0,0 +1,194 @@
+		error("Player_HE::loadAdLibBank(): could not open %s", drvName.c_str());
+
+	uint32 size = uint32(file.size());
+	
```
This line has trailing whitespaces

sev- commented on audio/mididrv.h:
```diff
@@ -213,6 +213,9 @@ class MidiDriver : public MidiDriver_BASE {
 	virtual void close() = 0;
 
+	/** Set global music volume */
+	virtual void setVolume(int volume) { }
```
This is hides overloaded functions at engines/sci/sound/drivers/amigamac.cpp, engines/parallaction/adlib.cpp, engines/agos/drivers/accolade/adlib.cpp, /engines/queen/midiadlib.h, engines/sherlock/scalpel/drivers/adlib.cpp which define it as setVolume(byte)

sev- commented on engines/scumm/music.h:
```diff
@@ -55,6 +55,13 @@ class MusicEngine {
+	 * @param sound		the sound to start
+	 * @param track		the track to start
+	 */
+	virtual void startSound(int sound, int track) { startSound(sound); }
```
This shadows overloaded startSound() in engines/scumm/imuse_digi/dimuse.h.

sev- commented:
I forgot to mention that the merge commit must be not part of this PR too. In order to avoid it, just rebase your patch.

sev- commented on audio/miles_adlib.cpp:
```diff
@@ -270,7 +270,7 @@ class MidiDriver_Miles_AdLib : public MidiDriver {
 
 MidiDriver_Miles_AdLib::MidiDriver_Miles_AdLib(InstrumentEntry *instrumentTablePtr, uint16 instrumentTableCount)
-	: _masterVolume(15), _opl(0),
+	: _masterVolume(255), _opl(0),
```
Were other engines which use this driver tested with this change? Why this change is required at all?

sev- commented:
Very good. Thank you a lot for the change. Merging now.

Also, if you're interested in working further on HE sounds, I have something to discuss on the IRC. Come, talk to me (nick _sev at #scummvm on freenode)

nukeykt commented:
I am busy with other projects now, but i'm still interested. I'll contact you when i have a spare time.
